int32 latent_x
int32 latent_y